Content.Server/Magic/MagicSystem.cs

```diff
@@ -0,0 +1,321 @@
+        {
+            var spell = new WorldTargetAction(_prototypeManager.Index<WorldTargetActionPrototype>(id));
+            _actionsSystem.SetCharges(spell, charges < 0 ? null : charges);
+            component.Spells.Add(spell);
+        }
+
+        foreach (var (id, charges) in component.InstantSpells)
+        {
+            var spell = new InstantAction(_prototypeManager.Index<InstantActionPrototype>(id));
+            _actionsSystem.SetCharges(spell, charges < 0 ? null : charges);
+            component.Spells.Add(spell);
+        }
+
+        foreach (var (id, charges) in component.EntitySpells)
+        {
+            var spell = new EntityTargetAction(_prototypeManager.Index<EntityTargetActionPrototype>(id));
+            _actionsSystem.SetCharges(spell, charges < 0 ? null : charges);
+            component.Spells.Add(spell);
+        }
+    }
+
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+
+        foreach (var forcewall in EntityQuery<ForceWallSpellComponent>())
+        {
+            forcewall.Lifetime -= frameTime;
+
+            if (forcewall.Lifetime <= 0)
+                EntityManager.QueueDeleteEntity(forcewall.Owner);
+        }
+
+        foreach (var spawnedEntities in EntityQuery<SpawnSpellComponent>())
+        {
+            spawnedEntities.Lifetime -= frameTime;
+
+            if (spawnedEntities.Lifetime <= 0)
+                EntityManager.QueueDeleteEntity(spawnedEntities.Owner);
+        }
+    }
+
+    private void OnUse(EntityUid uid, SpellbookComponent component, UseInHandEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        AttemptLearn(uid, component, args);
+
+        args.Handled = true;
+    }
+
+    private void AttemptLearn(EntityUid uid, SpellbookComponent component, UseInHandEvent args)
+    {
+        component.CancelToken = new();
+        var doAfterEventArgs = new DoAfterEventArgs(args.User, component.LearnTime, component.CancelToken.Token, uid)
+        {
+            BreakOnTargetMove = true,
+            BreakOnUserMove = true,
+            BreakOnDamage = true,
+            BreakOnStun = true,
+            NeedHand = true, //What, are you going to read with your eyes only??
+            TargetFinishedEvent = new LearnDoAfterComplete(args.User),
+            TargetCancelledEvent = new LearnDoAfterCancel(),
+        };
+
+        _doAfter.DoAfter(doAfterEventArgs);
+    }
+
+    private void TeachSpell(EntityUid uid, SpellbookComponent component, LearnDoAfterComplete ev)
+    {
+        component.CancelToken = null;
+
+        if (ev.User == null)
+            return;
+
+        _actionsSystem.AddActions(ev.User, component.Spells, uid);
+    }
+
+    #region Spells
+
+    private void OnRuneMagic(RuneMagicEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        var transform = Transform(args.Performer);
+        Spawn(args.RunePrototype, transform.Coordinates.SnapToGrid());
```

Comment:
Does this cause any problems with instant collision trigger runes? Would it be better to make it spawn on a clicked tile?